terraform/examples/gcp/nat_gateway/main.tf

```diff
@@ -0,0 +1,106 @@
+terraform {
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = "5.19.0"
+    }
+  }
+}
+
+provider "google" {
+  project = var.project_id
+  region  = var.region
+  zone    = var.zone
+}
+
+resource "google_service_account" "firezone" {
+  account_id   = "firezone-gateway"
+  display_name = "Firezone Gateway Service Account"
+}
+
+resource "google_compute_network" "firezone" {
+  name                    = "firezone-gateway"
+  auto_create_subnetworks = true
+}
+
+resource "google_compute_router" "firezone" {
+  name    = "firezone-gateway-router"
+  network = google_compute_network.firezone.id
+}
+
+resource "google_compute_router_nat" "firezone" {
+  name                               = "firezone-gateway-nat"
+  router                             = google_compute_router.firezone.name
+  nat_ip_allocate_option             = "AUTO_ONLY"
+  source_subnetwork_ip_ranges_to_nat = "ALL_SUBNETWORKS_ALL_IP_RANGES"
+}
+
+resource "google_compute_instance_template" "gateway" {
+  name                 = "gateway-template"
+  description          = "Instance template for the Firezone Gateway"
+  instance_description = "Firezone Gateway"
+  machine_type         = var.machine_type
+  tags                 = ["gateway"]
+  can_ip_forward       = true
+
+  scheduling {
+    automatic_restart   = true
+    on_host_maintenance = "MIGRATE"
+  }
+
+  disk {
+    source_image = "debian-cloud/debian-12"
+    auto_delete  = true
+    boot         = true
+  }
+
+  network_interface {
+    network = google_compute_network.firezone.id
+  }
+
+  service_account {
+    email  = google_service_account.firezone.email
+    scopes = ["cloud-platform"]
+  }
+
+}
+
+// Allow SSH access to the gateways. This is optional but helpful for debugging
+// and administration of the gateways. Since they're not publicly accessible,
+// you need to tunnel through IAP:
+//
+//   gcloud compute ssh --tunnel-through-iap --project <PROJECT_ID> --zone <ZONE> gateway-0
+resource "google_compute_firewall" "ssh-rule" {
+  name    = "allow-ssh"
+  network = google_compute_network.firezone.name
+
+  allow {
+    protocol = "tcp"
+    ports    = ["22"]
+  }
+
+  target_tags   = ["gateway"]
+  source_ranges = ["0.0.0.0/0"]
+}
+
+resource "google_compute_instance_from_template" "gateway" {
```

Comment:
> So everything we have in ### Minimal downtime upgrades doesn't make any sense to do.

If the input vars (token) aren't changed for an `instance_template` then terraform won't roll them over, at least based on my testing.

> Once #2465 is closed we should go and update everything to start using them.

Does that mean that issue is a blocker for merging this?

This is meant to serve as a basic example for a customer -- I think it's worth having basic examples for users who don't need minimal downtime deploys and would rather have something simple to get up and running with quickly.

Reply:
I updated this example with a `Future improvements` section that includes this and some customer feedback. Also removed the "Minimal downtime upgrades" section since that's not quite where we want to steer customers per your feedback.

Reply:
> If the input vars (token) aren't changed for an instance_template then terraform won't roll them over, at least based on my testing.

If we release a new version it will roll them over.

> Does that mean that issue is a blocker for merging this?

Nope, we still can have zero-downtime deployment using the proper boot timeouts. It's just not as reliable as health checks.

> This is meant to serve as a basic example for a customer -- I think it's worth having basic examples for users who don't need minimal downtime deploys and would rather have something simple to get up and running with quickly.

To use a regional instance group we only need to replace `instance_from_template` with another kind of resource so complexity-wise is about the same but the example will give a better way to deploy and manage replicated apps. Since it's already merged I can update the example later.